src/pybind/mgr/dashboard_v2/HACKING.rst

```diff
@@ -0,0 +1,458 @@
+    import { Router } from '@angular/router';
+
+    import { ToastsManager } from 'ng2-toastr';
+
+    import { Credentials } from '../../../shared/models/credentials.model';
+    import { HostService } from './services/host.service';
+
+
+Backend Development
+-------------------
+
+Unit Testing and Linting
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+We included a ``tox`` configuration file that will run the unit tests under
+Python 2 and 3, as well as linting tools to guarantee the uniformity of code.
+
+You need to install ``tox`` before running it. To install ``tox`` in your
+system, either install it via your operating system's package management
+tools, e.g. by running ``dnf install python3-tox`` on Fedora Linux.
+
+Alternatively, you can use Python's native package installation method::
+
+  $ pip install tox
+
+To run tox, run the following command in the root directory (where ``tox.ini``
+is located)::
+
+  $ tox
+
+If you just want to run a single tox environment, for instance only run the
+linting tools::
+
+  $ tox -e lint
+
+How to add a new controller?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If you want to add a new endpoint to the backend, you just need to add a
+class derived from ``BaseController`` decorated with ``ApiController`` in a
+Python file located under the ``controllers`` directory. The Dashboard module
+will automatically load your new controller upon start.
+
+For example create a file ``ping2.py`` under ``controllers`` directory with the
+following code::
+
+  import cherrypy
+  from ..tools import ApiController, BaseController
+
+  @ApiController('ping2')
+  class Ping2(BaseController):
+    @cherrypy.expose
+    def default(self, *args):
+      return "Hello"
+
+Every path given in the ``ApiController`` decorator will automatically be
+prefixed with ``api``. After reloading the Dashboard module you can access the
+above mentioned controller by pointing your browser to
+http://mgr_hostname:8080/api/ping2.
+
+It is also possible to have nested controllers. The ``RgwController`` uses
+this technique to make the daemons available through the URL
+http://mgr_hostname:8080/api/rgw/daemon::
+
+  @ApiController('rgw')
+  @AuthRequired()
+  class Rgw(RESTController):
+    pass
+
+
+  @ApiController('rgw/daemon')
+  @AuthRequired()
+  class RgwDaemon(RESTController):
+
+    def list(self):
+      pass
+
+
+Note that paths must be unique and that a path like ``rgw/daemon`` has to have
+a parent ``rgw``. Otherwise it won't work.
+
+How does the RESTController work?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+We also provide a simple mechanism to create REST based controllers using the
+``RESTController`` class. Any class which inherits from ``RESTController``
+will, by default, return JSON.
+
+The ``RESTController`` is basically an additional abstraction layer which eases
+and unifies the work with collections. A collection is just an array of
+objects with a specific type. ``RestController`` enable some default mappings
```

Comment:
Possibly:
* ``RestController`` should be ``RESTController``?
* And "enable some default mappings" should be "enables some default mappings"?

Reply:
Good spotting! Fixed.